packages/hardhat/contracts/LargeGrant.sol

```diff
@@ -0,0 +1,390 @@
+					number: grant.stages[i].milestones[j].number,
+					amount: grant.stages[i].milestones[j].amount,
+					completed: grant.stages[i].milestones[j].completed,
+					approvedBy: grant.stages[i].milestones[j].approvedBy,
+					completedBy: grant.stages[i].milestones[j].completedBy
+				});
+				index++;
+			}
+		}
+	}
+
+	function addGrant(
+		address _builder,
+		uint256 _grantId,
+		uint256[] memory _milestoneAmounts
+	) public onlyRole(OWNER_ROLE) {
+		GrantData storage grant = grants[_grantId];
+
+		if (grant.builder != address(0)) {
+			revert GrantAlreadyExists();
+		}
+
+		if (_builder == address(0)) {
+			revert ZeroAddress();
+		}
+
+		if (_milestoneAmounts.length == 0) {
+			revert NoMilestones();
+		}
+
+		grant.builder = _builder;
+
+		Stage storage stage = grant.stages.push();
+		stage.number = 1;
+
+		for (uint i = 0; i < _milestoneAmounts.length; i++) {
+			uint256 amount = _milestoneAmounts[i];
+			if (amount == 0) {
+				revert MilestoneZeroAmount();
+			}
+			stage.milestones.push(
+				Milestone({
+					number: uint8(i + 1),
+					amount: amount,
+					completed: false,
+					approvedBy: address(0),
+					completedBy: address(0)
+				})
+			);
+		}
+
+		emit AddGrant(_grantId, _builder, _milestoneAmounts);
+	}
+
+	function addGrantStage(
+		uint256 _grantId,
+		uint256[] memory _milestoneAmounts
+	) public onlyRole(OWNER_ROLE) {
+		GrantData storage grant = grants[_grantId];
+
+		if (grant.builder == address(0)) {
+			revert GrantDoesNotExist();
+		}
+
+		if (_milestoneAmounts.length == 0) {
+			revert NoMilestones();
+		}
+
+		Stage storage stage = grant.stages.push();
+
+		stage.number = uint8(grant.stages.length);
+
+		for (uint i = 0; i < _milestoneAmounts.length; i++) {
+			uint256 amount = _milestoneAmounts[i];
+			if (amount == 0) {
+				revert MilestoneZeroAmount();
+			}
+			stage.milestones.push(
+				Milestone({
+					number: uint8(i + 1),
+					amount: amount,
+					completed: false,
+					approvedBy: address(0),
+					completedBy: address(0)
+				})
+			);
+		}
+
+		emit AddGrantStage(
+			_grantId,
+			uint8(grant.stages.length),
+			_milestoneAmounts
+		);
+	}
+
+	function approveMilestone(
+		uint256 _grantId,
+		uint8 _stageNumber,
+		uint8 _milestoneNumber
+	)
+		public
+		whenNotPaused
+		onlyRole(OWNER_ROLE)
+		validMilestone(_grantId, _stageNumber, _milestoneNumber)
+	{
+		GrantData storage grant = grants[_grantId];
+
+		if (
+			grant
+				.stages[_stageNumber - 1]
+				.milestones[_milestoneNumber - 1]
+				.approvedBy != address(0)
+		) {
+			revert MilestoneAlreadyApproved();
+		}
+
+		grant
+			.stages[_stageNumber - 1]
+			.milestones[_milestoneNumber - 1]
+			.approvedBy = msg.sender;
+
+		emit ApproveMilestone(
+			_grantId,
+			_stageNumber,
+			_milestoneNumber,
+			msg.sender
+		);
+	}
+
+	function completeMilestone(
+		uint256 _grantId,
+		uint8 _stageNumber,
+		uint8 _milestoneNumber,
+		string memory _description,
+		string memory _proof
+	)
+		public
+		whenNotPaused
+		onlyRole(OWNER_ROLE)
```

Comment:
> Added milestones approval

Umm I think we haven't implemented multiple approval yet right? This will allow owner / stweard to approve the grant?

Reply:
Yes, one owner/steward needs to approve the milestone and then another one can complete the milestone, sending the funds.

The grants are approved off-chain before one owner/steward can create the grant on-chain.